core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -987,6 +987,25 @@ default ConnectorTableProperties getTableProperties(ConnectorSession session, Co
         return Optional.empty();
     }
 
+    /**
+     * Attempt to push down the join operation.
+     * <p>
+     * Connectors can indicate whether they don't support join pushdown or that the action had no effect
+     * by returning {@link Optional#empty()}. Connectors should expect this method may be called multiple times.
+     * </p>
+     */
+    default Optional<ConnectorTableHandle> applyJoin(
+            ConnectorSession session,
+            JoinType joinType,
+            ConnectorTableHandle left,
+            ConnectorTableHandle right,
+            ConnectorExpression joinCondition,
+            Map<String, ColumnHandle> leftAssignments,
+            Map<String, ColumnHandle> rightAssignments)
```

Comment:
I think we need assignments separately. column handles from two different table scans are not necessarily distinguishable (often they are not in practice)

Reply:
If the are not distinguishable then `joinCondition` might be not solvable.

Reply:
Tha's why i have assignments as separate bags

Reply:
@sopel39 consider `orders o1 join orders o2 on o1.orderkey < o2.orderkey`
then the `joinCondition` will be a connector expression representing `orderkey_1 < orderkey_2` 
and assignments will be `orderkey_1 => ColumnHandle(orderkey)`, `orderkey_2 => ColumnHandle(orderkey)`.
The connector needs to distinguish which column handle is coming from the left and which from the right relation.

Reply:
> The connector needs to distinguish which column handle is coming from the left and which from the right relation.

It should be able to do so from `joinCondition`, no? Why to communicate that twice? Can connector expect that `joinConditions` will be flipped regarding left/right symbols? This might be confusing